### PR TITLE
[onnx-to-tosa] FXML-1101 Convert AveragePool ONNX To TOSA

### PR DESCRIPTION
**Pull request by maxbartel**

### CODE REVIEW

File: src/Conversion/ONNXToTOSA/ONNXToTOSACommon.cpp [hunk 1]
```diff
@@ -429,5 +429,6 @@ llvm::Optional<Value> convertReduceMeanOp(PatternRewriter &rewriter,
 
   return val;
 }
+
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/ONNXToTOSACommon.cpp:432]: super nit: Unnecessary change

File: src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp [hunk 2]
```diff
@@ -38,13 +39,102 @@ namespace tosa {
...
+  input = tosaBuilder.transpose(input, {0, 2, 3, 1});
+
+  // When ceil mode is 1, we need to add values to the padding
+  const llvm::SmallVector<int64_t, 4> ceilConstants =
+      getCeilConstants<ONNXPoolOp>(inputType.getShape(), shapeHelper, ceilMode);
+  llvm::SmallVector<int64_t, 4> pads =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // reorder padding values
+  mlir::ArrayAttr newPads = rewriter.getI64ArrayAttr({pads[0],
```
Comment (flemairen6) [inline src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp:124]: Nit : This will create a I64 const, and the previous call to tosaBuild.transpose actually creates a i32 const if my memory is correct. I noticed in the IR recently that we had a mix of both for constants. It's not issue, but that would be great if we we always using the same type. What do you think?

Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp:124]: You are right, I changed most if the implementation to i32, because that is what the TOSA Specification says. TOSA MLIR allows a bit more. I will change this to i32 to reflect the specification!

File: src/Conversion/ONNXToTOSA/NN/AveragePool.cpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+
+    Value input = adaptor.X();
+    // The attributes storage_order and dilations are unsupported
+    auto inputType = input.getType().cast<TensorType>();
+    const int64_t includePad = adaptor.count_include_pad();
+
+    if (inputType.getShape().size() != 4) {
+      return rewriter.notifyMatchFailure(
+          averagePoolOp, "TOSA only supports maxpool 2d");
```
Comment (ljfitz) [inline src/Conversion/ONNXToTOSA/NN/AveragePool.cpp:48]: maxpool? Here and a couple of other places. Could the various poolings be reasonably templated?

File: src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp (hunk 2 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp:124]: Interesting enough, TOSA MLIR wants this to be i64 Attributes. Seems to be a departure from the standard. I will not change this then, but it is indeed strange

File: src/Conversion/ONNXToTOSA/NN/AveragePool.cpp (hunk 3 above)
Comment (maxbartel) [reply src/Conversion/ONNXToTOSA/NN/AveragePool.cpp]: I changed this check to the common function and it now says `"TOSA only supports 2d pooling"`. The checks that remain in the original `matchAndRewrite `are not shared across different ops.

**Review by ljfitz — APPROVED** [review]
